lua/entities/magicbutton.lua

```diff
@@ -0,0 +1,463 @@
+            MAGICBUTTON_ENT_TABLE[ent] = nil
+        end
+    end)
+
+    hook.Add("OnEntityCreated", "magicbutton_ent_register", function(ent)
+        if (ent:GetClass() == "magicbutton") then
+            MAGICBUTTON_ENT_TABLE[ent] = true
+        end
+    end)
+end
+
+local MAGICBUTTON_HULLSIZE = Vector(32, 32, 72) --Vector(32,32,36)
+
+function ENT:IsTraceValid(trace, final)
+    local surprop = util.GetSurfacePropName(trace.SurfaceProps)
+    if (table.HasValue(self.PlacementSettings.SurfacePropBlacklist, surprop)) then return false end
+    if (table.HasValue(self.PlacementSettings.TextureBlacklist, trace.HitTexture)) then return false end
+    if (not trace.Hit) then return false end
+    if (trace.HitSky) then return false end
+    if (trace.HitNoDraw) then return false end
+    if (IsValid(trace.Entity)) then return false end
+    if (trace.AllSolid) then return false end
+    if (trace.StartSolid and trace.FractionLeftSolid == 0) then return false end
+
+    if (not self.PlacementSettings.AllowFloor) then
+        if (final and trace.HitNormal.z > 0.8) then return false end
+    end
+
+    return true
+end
+
+--MAGICBUTTON_CACHED_ORIGINPOINTS = nil
+function ENT:FindSuitableCastOrigin()
+    local hull = MAGICBUTTON_HULLSIZE
+    if (IsValid(self.Owner)) then end --return self:GetPos() + Vector(0,0,64)
+    local navareas = navmesh.GetAllNavAreas()
+    local picks
+
+    if (MAGICBUTTON_CACHED_ORIGINPOINTS == nil) then
+        local origins = {}
+
+        for _, area in pairs(navareas) do
+            local extent = area:GetExtentInfo()
+            local toosmall = extent.SizeX < 32 or extent.SizeY < 32
+            if (area:IsUnderwater() or area:HasAttributes(NAV_MESH_AVOID) or toosmall) then continue end
+            local tr = {}
+            tr.start = area:GetCenter() + Vector(0, 0, hull.z / 4)
+            tr.endpos = tr.start + Vector(0, 0, -64)
+            tr.mask = MASK_PLAYERSOLID
+            tr.mins = hull * Vector(1, 1, 0.1) * -0.5
+            tr.maxs = hull * Vector(1, 1, 0.1) * 0.5
+            local trace = util.TraceHull(tr)
+            if (not self:IsTraceValid(trace)) then continue end
+            local tr = {}
+            tr.start = trace.HitPos
+            tr.endpos = tr.start + Vector(0, 0, self.PlacementSettings.CeilingHeight)
+            tr.mask = MASK_PLAYERSOLID
+            tr.mins = hull * Vector(1, 1, 0.1) * -0.5
+            tr.maxs = hull * Vector(1, 1, 0.1) * 0.5
+            trace = util.TraceHull(tr)
+
+            if ((not self.PlacementSettings.RequireCeiling or (trace.Hit and self:IsTraceValid(trace))) and not trace.StartSolid or (trace.StartSolid and trace.FractionLeftSolid < 1)) then
+                table.insert(origins, trace)
+                local ext = area:GetExtentInfo()
+                local bnds = Vector(ext.SizeX, ext.SizeY, ext.SizeZ + 1)
+                debugoverlay.Box(area:GetCenter() + Vector(0, 0, 0.5), bnds * -0.5, bnds * 0.5, 60, Color(0, 0, 255, 0))
```

Comment:
debug overlay?

Reply:
These are debugging functions used for visualizing. They apparently do nothing unless developer is set to 1.